File: tcp_server/tcp_server_basic.cpp
// As written by copilot


#include <iostream>
#include <cstring>
#include <unistd.h>
#include <sys/socket.h>
#include <netinet/in.h>

int main() {
    int server_fd, new_socket;
    struct sockaddr_in address;
    int opt = 1;
    int addrlen = sizeof(address);
    char buffer[1024] = {0};

    // Create socket
    if ((server_fd = socket(AF_INET, SOCK_STREAM, 0)) == 0) {
        perror("Socket creation failed");
        return 1;
    }

    // Set socket options
    if (setsockopt(server_fd, SOL_SOCKET, SO_REUSEADDR | SO_REUSEPORT, &opt, sizeof(opt))) {
        perror("Setsockopt failed");
        return 1;
    }

    address.sin_family = AF_INET;
    address.sin_addr.s_addr = INADDR_ANY;
    address.sin_port = htons(12345); // Change this to your desired port

    // Bind socket to address
    if (bind(server_fd, (struct sockaddr *)&address, sizeof(address)) < 0) {
        perror("Bind failed");
        return 1;
    }

    // Listen for incoming connections
    if (listen(server_fd, 3) < 0) {
        perror("Listen failed");
        return 1;
    }

    std::cout << "Listening on port 12345..." << std::endl;

    while (true) {
        if ((new_socket = accept(server_fd, (struct sockaddr *)&address, (socklen_t *)&addrlen)) < 0) {
            perror("Accept failed");
            return 1;
        }

        // Read data from client
        int bytes_received = read(new_socket, buffer, sizeof(buffer));
        if (bytes_received > 0) {
            std::cout << "Received: " << buffer << std::endl;
            // Echo back to client
            send(new_socket, buffer, bytes_received, 0);
        }

        close(new_socket);
    }

    return 0;
}

File: tcp_server/tcp_server_basic.py
#!/usr/bin/python3

import socket
import sys

def main():
    host = '127.0.0.1'  # Change this to your desired IP address
    port = 12345        # Change this to your desired port number

    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as server_socket:
        server_socket.bind((host, port))
        server_socket.listen()

        print(f"Listening on {host}:{port}")

        try:
            while True:
                client_socket, client_address = server_socket.accept()
                print(f"Connection from {client_address}")

                with client_socket:
                    while True:
                        data = client_socket.recv(1024)
                        if not data:
                            print("Client Disconnected")
                            break;
                        send_data = b"Hello from Server: " + data
                        print(f"Received msg:{data}")
                        print(f"sending:{send_data}")
                        client_socket.sendall(send_data)
        except KeyboardInterrupt:
            print("\nServer shutting down")
            sys.exit(0)

if __name__ == "__main__":
    main()

